instrumentation/netty-codec-http/src/main/java/brave/netty/http/TracingHttpServerHandler.java

```diff
@@ -0,0 +1,61 @@
+package brave.netty.http;
+
+import brave.Span;
+import brave.Tracer;
+import brave.Tracer.SpanInScope;
+import brave.http.HttpServerHandler;
+import brave.http.HttpTracing;
+import brave.propagation.TraceContext;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+
+final class TracingHttpServerHandler extends ChannelDuplexHandler {
+  final Tracer tracer;
+  final HttpServerHandler<HttpRequest, HttpResponse> handler;
+  final TraceContext.Extractor<HttpHeaders> extractor;
+
+  TracingHttpServerHandler(HttpTracing httpTracing) {
+    tracer = httpTracing.tracing().tracer();
+    handler = HttpServerHandler.create(httpTracing, new HttpNettyAdapter());
+    extractor = httpTracing.tracing().propagation().extractor(HttpHeaders::get);
+  }
+
+  @Override public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+    if (!(msg instanceof HttpRequest)) return;
+
+    HttpRequest httpRequest = (HttpRequest) msg;
+    Span span = handler.handleReceive(extractor, httpRequest.headers(), httpRequest);
+    ctx.channel().attr(NettyHttpTracing.SPAN_ATTRIBUTE).set(span);
+    // TODO figure out how to get the client IP from ctx and set it if not span.isNoop
+
+    try (SpanInScope ws = tracer.withSpanInScope(span)) {
+      super.channelRead(ctx, msg);
+    } catch (Exception | Error e) {
+      ctx.channel().attr(NettyHttpTracing.SPAN_ATTRIBUTE).set(null);
+      handler.handleSend(null, e, span);
+      throw e;
+    }
+  }
+
+  @Override
+  public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise prm) throws Exception {
+    if (!(msg instanceof HttpResponse)) return;
+
```

Comment:
You also want to call ctx.write(msg, promise) before you return as otherwise you will never complete the promise. Or if you want to fail it just do ao

Reply:
Great catch.. wonder how i can test this such that it will fail (ex using a client against a server with this bug)? Any advice?

Reply:
nevermind again :)